sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnInvokers.java

```diff
@@ -0,0 +1,631 @@
+      Class<? extends DoFn> clazz) {
+    final TypeDescription clazzDescription = new TypeDescription.ForLoadedType(clazz);
+
+    DoFnSignature signature = DoFnReflector.getSignature(clazz);
+
+    DynamicType.Builder<?> builder =
+        new ByteBuddy()
+            // Create subclasses inside the target class, to have access to
+            // private and package-private bits
+            .with(
+                new NamingStrategy.SuffixingRandom("auxiliary") {
+                  @Override
+                  public String subclass(TypeDescription.Generic superClass) {
+                    return super.name(clazzDescription);
+                  }
+                })
+            // Create a subclass of DoFnInvoker
+            .subclass(DoFnInvoker.class, ConstructorStrategy.Default.NO_CONSTRUCTORS)
+            .defineField(
+                FN_DELEGATE_FIELD_NAME, clazz, Visibility.PRIVATE, FieldManifestation.FINAL)
+            .defineConstructor(Visibility.PUBLIC)
+            .withParameter(clazz)
+            .intercept(new InvokerConstructor())
+            // Delegate processElement(), startBundle() and finishBundle() to the fn.
+            .method(ElementMatchers.named("invokeProcessElement"))
+            .intercept(new ProcessElementDelegation(signature.getProcessElement()))
+            .method(ElementMatchers.named("invokeStartBundle"))
+            .intercept(new BundleMethodDelegation(signature.getStartBundle()))
+            .method(ElementMatchers.named("invokeFinishBundle"))
+            .intercept(new BundleMethodDelegation(signature.getFinishBundle()));
+
+    DynamicType.Unloaded<?> unloaded = builder.make();
+
+    @SuppressWarnings("unchecked")
+    Class<? extends DoFnInvoker<?, ?>> res =
+        (Class<? extends DoFnInvoker<?, ?>>)
+            unloaded
+                .load(DoFnInvokers.class.getClassLoader(), ClassLoadingStrategy.Default.INJECTION)
+                .getLoaded();
+    return res;
+  }
+
+  private abstract static class MethodDelegation implements Implementation {
+    FieldDescription delegateField;
+
+    @Override
+    public InstrumentedType prepare(InstrumentedType instrumentedType) {
+      // Remember the field description of the instrumented type.
+      delegateField =
+          instrumentedType
+              .getDeclaredFields()
+              .filter(ElementMatchers.named(FN_DELEGATE_FIELD_NAME))
+              .getOnly();
+
+      // Delegating the method call doesn't require any changes to the instrumented type.
+      return instrumentedType;
+    }
+
+    @Override
+    public ByteCodeAppender appender(final Target implementationTarget) {
+      return new ByteCodeAppender() {
+        @Override
+        public Size apply(
+            MethodVisitor methodVisitor,
+            Context implementationContext,
+            MethodDescription instrumentedMethod) {
+          StackManipulation manipulation =
+              isRequired()
```

Comment:
This `isRequired` seems a bit hacky. What if we had a `NoopMethodDelegation`. Then the decision about whether to use `NoopMethodDelegation`, `ProcessElementDelegation`, etc. can be made earlier (based on whether or not there is a method to delegate to).